pkg/queue/breaker.go

```diff
@@ -170,52 +168,68 @@ func (s *Semaphore) Release() error {
 	}
 }
 
-// AddCapacity increases the number of tokens in the rotation.
-// Increases only if the size is positive, does nothing otherwise.
-// An error is returned if not all capacity could be added to the semaphore.
-// This error could happen when the number of tokens exceeds the semaphore's queue size.
-func (s *Semaphore) AddCapacity(size int32) error {
+// UpdateCapacity updates the capacity of the semaphore to the desired
+// size.
+func (s *Semaphore) UpdateCapacity(size int32) error {
+	if size < 0 {
+		return ErrReduceCapacity
+	}
+
 	s.mux.Lock()
 	defer s.mux.Unlock()
 
-	for i := int32(0); i < size; i++ {
-		select {
-		case s.queue <- s.token:
-			s.capacity++
-		default:
-			return ErrAddCapacity
-		}
+	if s.effectiveCapacity() == size {
+		return nil
 	}
-	return nil
-}
 
-// ReduceCapacity removes tokens from the rotation.
-// It tries to acquire as many tokens as possible. If there are not enough tokens in the queue,
-// it postpones the operation for the future by increasing the `reducers` counter.
-// We return an error when attempting to reduce more capacity than was originally added.
-func (s *Semaphore) ReduceCapacity(size int32) error {
-	s.mux.Lock()
-	defer s.mux.Unlock()
+	if size > s.maxCapacity {
+		return ErrAddCapacity
+	}
 
-	if size+s.reducers > s.capacity {
-		return ErrReduceCapacity
+	// Add capacity until we reach size, potentially consuming
+	// outstanding reducers first.
+	for s.effectiveCapacity() < size {
```

Comment:
May be we can just reduce number of reducers first and then loop to just put tokens in the queue? 
I like the elegant approach you did here, though.

Reply:
You end up calculating a lot of intermediary values and in the end that looks harder to understand than this block right now.